File: lecture_33/p2.cpp
#include<iostream>
using namespace std;

TreeNode* deleteNode(TreeNode* node, int key) {
    if(node==nullptr)
        return nullptr;
    if(node->val==key)
    {
        if(node->left==nullptr && node->right==nullptr)
            return nullptr;
        else if(node->left==nullptr)
            return node->right;
        else if(node->right==nullptr)
            return node->left;
        else
        {
            TreeNode* cur=node->left;
            while(cur->right!=nullptr)
                cur=cur->right;
            node->val=cur->val;
            node->left=deleteNode(node->left,cur->val);
        }
        return node;
    }
    else if(key<node->val)
        node->left=deleteNode(node->left,key);
    else
        node->right=deleteNode(node->right,key);
    return node;
}
int main(int args,char argv)
{
    return 0;
}

File: lecture_34/p2.java
public class p2 {
    public boolean solve(TreeNode node,long min,long max)
    {
        if(node==null)
            return true;
        if(node.val<min || node.val>max)
            return false;
        boolean l=solve(node.left,min,(long)node.val-1);
        boolean r=solve(node.right,(long)node.val+1,max);
        return (l&&r);
    }
    public boolean isValidBST(TreeNode root) {
        return solve(root,Integer.MIN_VALUE,Integer.MAX_VALUE);
    }
    // long is used because of this case
    // [-2147483648,-2147483648]
    public static void main(String[] args) {
    }
}
